Make ConditionFirstBoot safe against power failures

Rahix opened the pull request:
This is an attempt at implementing @poettering's proposal from issue #4511:

> Here's an idea, maybe on the first boot we should not write a new uuid to /etc/, but instead just an empty file, and then overmount that with a symlink to /run/machine-id, like we already do if /etc is not writable and the file is exists and is empty. Then, rely on systemd-machine-id-commit.service to write it to disk in later boot. Then, let's consider all cases where /etc/machine-id exists but is empty a first case boot oot. This should be a relatively easy change, and would mean until systemd-machine-id-commit.service has once been reached all boots are considered first boot.

The behavioral changes this would incur:

1. An empty `/etc/machine-id` is treated like a missing `/etc/machine-id`: `m->first_boot` will be set and thus `ConditionFirstBoot` units will run.
2. `systemd-machine-id-commit.service` runs later, after `basic.target` (instead of before `sysinit.target`).  This change is necessary to make sure any units with `ConditionFirstBoot=yes` have a chance to run _before_ the machine-id is committed.

I've tested these changes to the best of my abilities and everything seems to behave as I would expect.  One thing I'm a bit unsure about is whether `1.` will lead to issues anywhere.  On systems with a read-only rootfs, an empty `/etc/machine-id` needs to exist for systemd to boot so `ConditionFirstBoot` would now always be true for them.  I'm not sure if this is a problem?

Rahix commented:
I have some trouble understanding the CI failures ... Can someone give me a hint what is going wrong?  Is it the `proc-sys-fs-binfmt_misc.mount` unit failing?

mrc0mmand commented:
> I have some trouble understanding the CI failures ... Can someone give me a hint what is going wrong? Is it the `proc-sys-fs-binfmt_misc.mount` unit failing?

Yes. The *unprivileged systemd-nspawn* part of the `TEST-01-BASIC` fails with: 

```
[  445.090286] systemd-journald[16]: Journal started
[  445.090415] systemd-journald[16]: Runtime Journal (/run/log/journal/3d0dd55f68524906b922563e883b4d00) is 8.0M, max 159.1M, 151.1M free.
[  445.091277] systemd[15]: proc-sys-fs-binfmt_misc.mount: Executing: /usr/bin/mount binfmt_misc /proc/sys/fs/binfmt_misc -t binfmt_misc -o nosuid,nodev,noexec
[  445.091580] mount[15]: mount: /proc/sys/fs/binfmt_misc: permission denied.
```

poettering commented on src/core/machine-id-setup.c:
```diff
@@ -165,7 +167,8 @@ int machine_id_setup(const char *root, sd_id128_t machine_id, sd_id128_t *ret) {
 
-        log_info("Installed transient %s file.", etc_machine_id);
+        if (!always_transient)
+                log_info("Installed transient %s file.", etc_machine_id);
```
why not leave the log message in?

Rahix replied:
My thinking was that it does not provide value in the case where it will always be installed transiently.  But if you prefer to have it log always, I'll change it.

poettering replied:
a compromise might be:

```c
log_full(always_transient ? LOG_DEBUG : LOG_INFO, "…
```

poettering commented on src/core/machine-id-setup.c:
```diff
@@ -85,7 +85,7 @@ static int generate_machine_id(const char *root, sd_id128_t *ret) {
 }
 
-int machine_id_setup(const char *root, sd_id128_t machine_id, sd_id128_t *ret) {
+int machine_id_setup(const char *root, bool always_transient, sd_id128_t machine_id, sd_id128_t *ret) {
```
maybe call this force_transient?

poettering commented on units/systemd-machine-id-commit.service:
```diff
@@ -12,8 +12,8 @@ Description=Commit a transient machine-id on disk
-Before=sysinit.target shutdown.target
-After=local-fs.target
+Before=shutdown.target
+After=basic.target
```
So I can see why dropping the before= line makes sense. But I am not convinced about the After= change. I think we should still commit the uuid as early as possible, and not needlessly delay that. I see no reason to say that basic.target was a requirement to say that first boot was complete now. I mean, I think in many cases people might very well declare that running the preset stuff during early boot fully would be enough to not require us to re-run first boot mode. Hence running the service as early as possible should be OK.

Or to say this differently: I think OS builders should think about where to put the boundary for this, and explicitly order all services they want to finish fist before this boundary. I see no reason to do that for them.

By dropping the Before= line people can order both early-boot services before and regular services before this commit point, which makes a lot of sense. But other than that I'd not make a change here.

Rahix replied:
Okay, I was under the assumption that `m->first_boot` will be reset once basic.target is reached but upon reading the code again, it seems this happens after all jobs of the initial transaction completed?

What you're saying makes a lot of sense, I agree that committing should happen as early as possible and allowing regular services to be a part of the first boot is probably a nice feature as well.  Though, maybe a `first-boot.target` for the purposes of ordering makes more sense then?

poettering replied:
yes, i came to the same conclusion that f-b.t would make sense. see below.

poettering reviewed (changes requested):
Needs some docs (not sure where best, maybe in the systemd-machine-id-commit.service man page, and/or where ConditionFirstBoot= is documented).

poettering commented:
Hmm, so I think to make this fully safe src/machine-id-setup/machine-id-setup-main.c() should invoke sync() before it does what it does, so that all changes done before the service is invoked definitely have hit the disk before we drop the first boot state.

poettering commented:
Should carry "Fixes: #4511" in some commit, so that that issue is automatically closed by github when merged

poettering commented:
/cc @boucman

poettering commented:
/cc @utezduyar

Rahix commented on src/core/main.c:
```diff
@@ -2004,15 +2004,16 @@ static void log_execution_mode(bool *ret_first_boot) {
+                         * manager wants to provision the machine ID itself it should pass $container_uuid to
+                         * PID 1. */
+
+                        *ret_first_boot = !!missing_or_empty_path("/etc/machine-id");
```
Yeah, you were actually too fast for me to write a comment about it:  I think my current implementation is broken because of this.  Apparently most distros ship an empty /etc/machine-id which leads to systemd (with my changes) presetting all units.  That has caused the integration tests in CI to fail and would probably cause a lot of problems further down the line (from reading #8268).

So my idea was exactly what you describe: A 'uninitialized' value for the machine-id until the first boot succeeds.

---

That said, I think the CI failures might have made an unrelated problem surface:  Currently `proc-sys-fs-binfmt_misc.mount` would be preset to enabled which doesn't seem in line with its documentation and, as I observed, leads to various issues.  The temporary commit (e25743ff423bb2e98228a6e117e1a74f8c3515e3) I added seems to fix it.  What do you think?  Maybe I should open a separate PR for this?

poettering replied:
Hmm, sounds OK to include it here in a separate commit. (Also see 508133917d9628f8c0adc5db34d595196b72b1c8)

And yeah, I think it should be disabled by default. /cc @keszybz

keszybz replied:
> Hmm, sounds OK to include it here in a separate commit. (Also see 5081339)

+1



> * if /etc/machine-id exists and is initialized → all good, we do nothing
> 
> * if /etc/machine-id is missing → we initialize it and declare first boot mode
> 
> * if /etc/machine-id exists and is empty → we initialize/overmount it, but do not enter first boot mode
> 
> * if /etc/machine-id exists and contains special string "uninitialized" → we initialize/overmount it and declare first boot mode

Shouldn't the last two be reversed for backwards compatibility? IIUC, currently we enter first boot mode when /etc/machine-id exists and is empty. Instead of redefining this, I'd define a special value that means "use a temporary value on each boot". Maybe even `0000000000000000`.

Rahix replied:
@keszybz, if /etc/machine-id exists but is empty, this is currently **not** considered a first boot and the idea was to keep this as is (for backwards compatibility).  So there is no redefinition of any exisiting value here.

Rahix commented:
> Needs some docs (not sure where best, maybe in the systemd-machine-id-commit.service man page, and/or where ConditionFirstBoot= is documented).

ACK, I just wanted to wait with documentation until it's clear how the mechanism should work.

poettering commented:
And I guess the man page machine-id(5) should also carry information about the new semantics.

Hmm, and I think `systemd-random-seed.service` should be updated to be ordered before `systemd-machine-id.service`. It ignores the saved random seed during first boot, but still updates it with a new one. We need to make sure that this has completed before the subsequent boots are started where the random seed is then used. or in other words: as long as the random seed hasn't been initialized once it should not be used.

Also, `systemd-firstboot.serivce` should be ordered before `systemd-machine-id.service`, so that it is triggred as long as the first boot is not complete.

poettering commented:
Hmm, thinking about this, one more idea: I think we should insert a generic target between the relevant services. i.e. introduce first-boot-complete.target or so. Which is the generic place where systemd-random-seed.service and systemd-firstboot.service order themselves before, and where systemd-machine-id-commit.service is ordered after.

Rahix commented:
Alright, I think I've got all things sorted out now.  Here's the changes:

- [x] I've added the `"uninitialized"` value for the machine-id.  It is written during early boot and (in most cases) overwritten with the machine-id later.  If the boot is aborted in between, the next boot will treat machine-id containing `uninitialized` like the file was missing -> boot is also considered first boot (4b33bbc0929f47ec2c979671b932267f066bf351, 34e953ea0cb887842919642285fed1ac2b590841).
- [x] Added the commit to ensure `proc-sys-fs-binfmt_misc.mount` is not preset to enabled (ac0c33d973b79377f6f09a3dc99c06443155e154).
- [x] Log message change as suggested.
- [x] Renamed `always_transient` to `force_transient` (though I'm not sure if the name still makes sense ... maybe `late_persist` or something similar would be better now?).
- [x] I've added the `first-boot.target` unit for ordering (not sure whether the way I added it into the dependency tree is good?) (30de32f67367de9729d6b0e1475bc398d3fc5976).
- [x] `sync()` call before finally committing the machine-id to disk (9e00ee9da73a7cc3d0b814df4ed8ba6ca9b11fc5).
- [x] Ordered `systemd-random-seed.service` and `systemd-firstboot.service` before `first-boot.target` (320edf2485569663157c53dc16aaf56f05a52fad, 6e862527251fcd02c77b0f04168fb4dceafabaf5).
- [x] Documentation for the new (and existing) machine-id semantics (6383d4420058e52795edad77f7369835652c633a).
- [x] I've checked (hopefully) all places where `/etc/machine-id` is read to make sure they will cope with the new marker value (3c1ef671db78c4f780e44449770c715e950542f5, eb4d313c731d9584d02d1c62d71001ad75ebcf47).

I've also built myself a small test-script to check all the new invariants.  It's looking good so far though I couldn't really test whether there are any races/non-atomic related issues remaining.

Rahix commented:
Added the Fixes: tag as well now ...

poettering commented:
Hmm, I wonder what the relationship between boot-complete.target and first-boot-complete.target should be. I.e. the former is for "boot counting/blessing", i.e. it's only reached after the point where we decided all is good so that the current boot menu item is marked as good forever. I wonder if one should be ordered after the other or pull it in.

I am a bit unsure, maybe they don't actually have a relationship, I am just wondering...

Rahix commented on src/core/main.c:
```diff
@@ -2004,15 +2004,24 @@ static void log_execution_mode(bool *ret_first_boot) {
+                                log_info("Running for the first time.");
+                        } else if (strcmp(id_text, "uninitialized") == 0) {
+                                *ret_first_boot = true;
+                                log_info("Running for the first time (detected previous incomplete boot).");
```
> I totally see usecases where one would initialize the file like this on a deployment image.

I also had this thought, but actually, I'm not sure where this would make sense.  `uninitialized` behaves as if the file was missing entirely, so any image which wants to have the first-boot detection should deploy without `/etc/machine-id`.

For a read-only rootfs it also doesn't make much sense, as this would mean systemd always thinks it's booting for the first time.

But I can remove the message, I agree that it might be better without it.

poettering replied:
but it would allow people to start with a read-only root, that is then remounted later on writable, and consider that a first boot scenario. That is very useful. After all systemd-machine-id-commit has this beautiful logic of commiting the machine ID atomically after the root disk has been mounted writable

Rahix commented on src/core/machine-id-setup.c:
```diff
@@ -142,13 +143,35 @@ int machine_id_setup(const char *root, sd_id128_t machine_id, sd_id128_t *ret) {
+
+                        r = fsync_directory_of_file(fd);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to sync directory of %s: %m", etc_machine_id);
```
> use fsync_full() instead.

Makes sense, thanks!

> Also I think we should do it one if level up, i.e. so that we also sync things when we properly initialize the machine ID

`id128_write_fd()` already calls fsync() and fsync_directory_of_file() so it's not necessary.

Rahix commented on src/core/machine-id-setup.c:
```diff
@@ -207,6 +207,9 @@ int machine_id_commit(const char *root) {
         int r;
 
+        /* Before doing anything, sync everything to ensure any changes by first-boot units are persisted */
+        sync();
```
I'm not sure how?  `sync()` doesn't return errors and I don't know any equivalent syscall that does ... Maybe I'm missing something?

poettering replied:
ah, sync() is stupid for not returning an error. Maybe use syncfs() on /etc and /var first, and then follow-up with sync(). After all this is mostly about stuff to change in /etc and /var, and what we mostly care about. i.e.:

```c
const char *path;

/* First, explicitly sync the file systems we care about and check if it worked */
FOREACH_STRING(path, "/etc/", "/var/") {
        r = syncfs_path(path);
        if (r < 0)
                return log_errno(r, "…");
}

/* Afterwards, sync() the rest too, but we can't check the return value for these */
sync();
```

That way we'll catch the errors we really care about, and the sync the rest as well. Not perfect but as close to perfect as it can be.

Rahix commented on units/first-boot.target:
```diff
@@ -0,0 +1,15 @@
+[Unit]
+Description=First boot completed
+Documentation=man:systemd.special(7)
+After=sysinit.target
```
I got a dependency cycle without it because `first-boot.target` is currently pulled in by `sysinit.target` ... But with the changes to the dependency structure, I guess this will no longer be necessary.

Rahix commented:
Okay, I'm still a bit unsure about the details of the unit structure.  From your comments, this is the overall picture I'm getting:

- `systemd-machine-id-commit.service` wants `first-boot-complete.target`
- `systemd-machine-id-commit.service` after `first-boot-complete.target`
- `systemd-machine-id-commit.service` wanted-by `sysinit.target`.

Then, units which should run on first boot have a `Before=first-boot-complete.target` so

- `systemd-firstboot.service` before `first-boot-complete.target`
- `systemd-random-seed.service` before `first-boot-complete.target`

Is this what you have in mind?

If I'm not mistaken, this would still allow people to abuse `first-boot-complete.target` as a hook by adding

```ini
[Install]
WantedBy=first-boot-complete.target
```

to their units.  Is there any way we can effectively prevent that?

Rahix commented:
> Hmm, I wonder what the relationship between boot-complete.target and first-boot-complete.target should be. I.e. the former is for "boot counting/blessing", i.e. it's only reached after the point where we decided all is good so that the current boot menu item is marked as good forever. I wonder if one should be ordered after the other or pull it in.
> 
> I am a bit unsure, maybe they don't actually have a relationship, I am just wondering...

Actually, this opens an interesting question: `boot-complete.target` works by adding `Requires=` dependencies on all units needed for the boot to complete.  If any of them fail, the boot is not blessed.

Maybe the same thing makes sense for `first-boot-complete.target`?  I.e. if any of the first boot jobs fail, we shouldn't persist the machine-id and give those units another try at the next boot?

I can see good points for and against this: E.g. a partitioning job definitely needs to complete for the system to work, but maybe some less important provisioning should not force the system to "first boot" again next time.  Especially, because the first boot presets all units which means as long as the job fails, effectively no manual changes to the enabled units can be done.

If we want to go that route, I would suggest adding `Requires=first-boot-complete.target` to `systemd-machine-id-commit.service` and allowing units to optionally add `RequiredBy=first-boot-complete.target` if they absolutely need to succeed.  Any other units will just run once, regardless of whether they succeed, and not influence whether the first boot will be marked as completed.

---

For the relationship between `boot-complete.target` and `first-boot-complete.target`, I think this largely depends on wether we do the above.  If yes, in theory adding

```
Requires=first-boot-complete.target
After=first-boot-complete.target
```

to `boot-complete.target` would make sense; if the first-boot jobs fail, the boot should not be blessed.  But OTOH, I'm a bit scared whether this would force something that might be considered distro policy.  After all, anyone could add those dependencies themselves if they need them.

poettering commented:
> Is this what you have in mind?

Yes.

> 
> If I'm not mistaken, this would still allow people to abuse `first-boot-complete.target` as a hook by adding
> 
> ```ini
> [Install]
> WantedBy=first-boot-complete.target
> ```
> 
> to their units. Is there any way we can effectively prevent that?

Well, we should document in systemd.special(7) how it is supposed to be used, but people can always shoot themselves in the foot if they really like and we shouldn't hinder them too hard from it.

poettering commented:
> I can see good points for and against this: E.g. a partitioning job definitely needs to complete for the system to work, but maybe some less important provisioning should not force the system to "first boot" again next time. Especially, because the first boot presets all units which means as long as the job fails, effectively no manual changes to the enabled units can be done.

Well, services should be able to choose if they want to hook themselves into the target via Requires= or not, i.e. whether they just want to delay the marking, or optionally block it entirely

keszybz commented on man/machine-id.xml:
```diff
@@ -115,6 +116,34 @@
+      writable).</para></listitem>
+
+      <listitem><para>If <filename>/etc/machine-id</filename> already contains a valid machine-id, this is
+      not a first boot.</para></listitem>
```
Hmm, does this preserve backwards compatibility or is it a change in behaviour?

poettering replied:
this is how it always was

poettering commented:
hmm, so, actually, thinking about this more: I think first-boot-complete.target and boot-complete.target probably should have a distinct nature:

1. f-b-c.t should be a synchronization point that is only pulled in by the stuff that wants to delay the "first boot done marking". This is what we typically call a "passive" unit, i.e. it doesn't pull anything on its own, it's sole role is to be a (potentially optional) sync point, that can be inserted if some service ordered before it.

2. b-c.t also should be a synchronization point, but it should be pulled in by the the stuff that wants to know when the boot is complete, and it pulls in what shall run before it. i.e. this is what we typically call an "active" unit, i.e. it actively pulls in services, that a consumer needs. And only if the consumer is enabled/used the target and its deps are pulled in.

Why does the difference matter? In the second case, the stuff ordered before the target is really stuff you don#t want unless you have boot blessing enabled. it's supposed to be stuff that checks the system for health and there's no point in invoking that if noone cares about the health state. In the first case we want to typically run the stuff before it in any case (i.e. random seed stuff always runs) but synced against machine-id-setup if so.

> > Maybe the same thing makes sense for `first-boot-complete.target`? I.e. if any of the first boot jobs fail, we shouldn't persist the machine-id and give those units another try at the next boot?
> 
> I think I agree with this.

So, given what I wrote above, I think I disagree afte rall: f-b-c.t should be fully passive, and thus not have Requires= line.

Rahix commented on src/libsystemd/sd-id128/id128-util.c:
```diff
@@ -97,6 +97,15 @@ int id128_read_fd(int fd, Id128Format f, sd_id128_t *ret) {
+        case 13:
+        case 14:
+                /* Treat an "uninitialized" id file like an empty one */
+                if (strncmp(buffer, "uninitialized\n", l) == 0)
```
> hmm, i don#t think we should do this unconditionally. It's too special. Let's enable this via some flag that can be ORed into the Id128Format type.

I'm not so sure whether this is a good idea.  I've implemented this to test how it would look and it doesn't feel right to me ... First of all, it would affect all in-tree places where `ID128_PLAIN` is currently used (short of the test-suite).  So the existence of `ID128_PLAIN` without the flag is only relevant to systemd API users at that point.

Secondly, I feel like this makes the mechanism a bit fragile.  People will easily forget this flag and have their code mistreat a machine-id from a system that crashed during first-boot.  As this is a case that will probably not see much test-coverage anyway, it could be hard to catch such incorrect code and nasty for those few who get to encounter it.  I myself wouldn't even have though about adding this extra case to `id128_read_fd()` if I didn't create a (local) test to try out an aborted boot (systemd-nspawn refusing to boot the image with "uninitialized" machine-id is how I found out).

As most code that reads `/etc/machine-id` will luckily use `id128_read()` or `id128_read_fd()`, having the special case handled there correctly all the time would prevent this accidental bug in most circumstances (the only manual parsing of `/etc/machine-id` is in `src/core/main.c` and `src/shared/dissect-image.c` which are both patched in this PR).

So I'd suggest keeping it like I implemented it right now.  I'd argue that it wouldn't really hurt other cases, as calling code already has to be robust against errors and whether `-EINVAL` or `-ENOMEDIUM` is returned in this one case shouldn't matter much.

poettering replied:
the thing is, we want in 99% of the cases that reading "uninitialized" from that file to be considered an error. After all we give the guarantee that /etc/machine-id is readable and initialized whenever any non-PID1 userspace is running. Thus, the only program that should ever see /etc/machine-id in this special state is PID 1 itself (and maybe systemd-machine-id-commit), all other tools should rely on it to be fully initiailzed

Rahix replied:
I fully agree with you that any tool/software running _in_ a system should be able to expect that `/etc/machine-id` is fully initialized.

However tools conceptually running _outside_ a system, that is tools operating on files in a system-image that is currently not running (let's call this a cold image) must be robust against finding `$ROOT/etc/machine-id` in the `uninitialized` state.  I've gone through all the relevant call-sites now to see where this could pose an issue.  Here's a summary:

### `src/core/machine-id-setup.c` - `machine_id_setup()`
:heavy_check_mark: Just needs to check whether there is a valid machine-id.  Returning any error here means a machine-id needs to be set up, whether that's `ENOENT`, `ENOMEDIUM`, `EINVAL` does not matter.

### `src/libsystemd/sd-id128/sd-id128.c` - `sd_id128_get_machine()`
:heavy_check_mark: This is meant to be called from within a running system to retrieve the machine-id.  Thus this will always see it in an initialized state.

### `src/machine-id-setup/machine-id-setup-main.c` - `run()`
:heavy_check_mark: Only relevant in the `if (arg_commit)` code path which will only run from within a system.

### `src/nspawn/nspawn.c` - `setup_machine_id()`
:x: nspawn reads `/etc/machine-id` of the cold image it is attempting to spawn.  Here we need to be prepared against finding it in `uninitialized` state, especially because nspawn makes a distinction between 'missing machine-id' (is okay) and 'invalid machine-id' (will refuse to spawn).  I've added commit 572e347c7c9f46786ffc0a747f13df4598725f62 to address this.

### `src/partition/repart.c` - `context_read_seed()`
:x: As far as I understood, repart usually runs from initramfs and reads out the machine-id of the cold, not-yet-booted rootfs.  Thus this is also a situation where `uninitialized` must be dealt with properly.  Fixed with commit 378c41ca4ae768b40d562ecaa44b3afac230566d.

### `src/shared/machine-image.c` - `image_read_metadata()`
:heavy_check_mark: While this function seems to operate on a cold image as far as I can tell, it does not discriminate between between errors returned by `id128_read_fd()`.  Thus I would expect this code is only called when the image is supposed to be in good shape already?  Or is this a potentially not yet updated place that does not properly handle empty (and now `uninitialized`) machine-id files?

### `src/test/test-fs-util.c` - `test_chase_symlinks()`
:heavy_check_mark: Just reads the machine-id from the currently running system.

---

Honestly, I still don't think this change is a good idea.  It is really easy for a future reader of a "cold machine-id" to mess this up without knowing about it.  Just searching and understanding all the call-sites just now was a cumbersome task and I'd not be 100% confident that I catched everything.

The already covered situation of `/etc/machine-id` being an empty file will return `-ENOMEDIUM` in all situations right now (and most code is set up to deal with this correctly).  As the `uninitialized` value is conceptually very similar, I do not really see why it should not be treated equally and has to require this additional explicitness at call-site (which is easy to forget/don't know about).

That said, I've now implemented the change as requested if you prefer it this way.

Rahix commented:
I've addressed a few of your review comments but for others I was a bit unsure; see my other answers.  Here's the changes I made since your last review:

- [x] Added a warning when reading `/etc/machine-id` fails with anything but `ENOENT`.
- [x] Fixed coding style issues and switched to `str{,n}eq` instead of `str{,n}cmp`.
- [x] Use `fsync_full()` instead of manually syncing.
- [x] Updated first-boot messages to better reflect their meaning.  Also added a `log_debug()` for the case when the boot is _not_ the first.
- [x] Renamed target to `first-boot-complete.target`.
- [x] Pull in `first-boot-complete.target` via `systemd-machine-id-commit.service` instead of `sysinit.target`.

Additionally, we've also hit an [interesting CI failure][1]: The test was finished so quickly that the system didn't have time to run `systemd-machine-id-commit.service` to completion.  This lead to `/etc/machine-id` containing "uninitialized" on the next boot -> we got a first-boot which broke the following test.  I've included a fix, ensuring that a non-first-boot system can't ever become first-boot in this way (when not first-boot the "uninitialized" value is never written).

[1]: https://jenkins-systemd.apps.ocp.ci.centos.org/job/upstream-vagrant-archlinux/46/changes

<details>
<summary>This is the fix's diff in isolation</summary>

```diff
diff --git a/src/core/main.c b/src/core/main.c
index e02546c339f2..e21a73318d8d 100644
--- a/src/core/main.c
+++ b/src/core/main.c
@@ -2040,6 +2040,7 @@ static void log_execution_mode(bool *ret_first_boot) {
 
 static int initialize_runtime(
                 bool skip_setup,
+                bool first_boot,
                 struct rlimit *saved_rlimit_nofile,
                 struct rlimit *saved_rlimit_memlock,
                 const char **ret_error_message) {
@@ -2073,7 +2074,8 @@ static int initialize_runtime(
 
                         status_welcome();
                         hostname_setup();
-                        machine_id_setup(NULL, true, arg_machine_id, NULL);
+                        /* Force transient machine-id on first boot. */
+                        machine_id_setup(NULL, first_boot, arg_machine_id, NULL);
                         (void) loopback_setup();
                         bump_unix_max_dgram_qlen();
                         bump_file_max_and_nr_open();
@@ -2801,6 +2803,7 @@ int main(int argc, char *argv[]) {
         log_execution_mode(&first_boot);
 
         r = initialize_runtime(skip_setup,
+                               first_boot,
                                &saved_rlimit_nofile,
                                &saved_rlimit_memlock,
                                &error_message);
```
</details>

---

The remaining open review comments are:

- Catch `sync()` failures
- Behavior in `id128-util`

poettering commented on src/libsystemd/sd-id128/id128-util.c:
```diff
@@ -97,6 +98,15 @@ int id128_read_fd(int fd, Id128Format f, sd_id128_t *ret) {
+                if (strneq(buffer, "uninitialized\n", l))
+                        return -ENOMEDIUM;
+                else
+                        return -EINVAL;
```
use ternary op?

poettering commented on man/systemd.special.xml:
```diff
@@ -277,6 +278,16 @@
+          <listitem>
+            <para>This target is intended as a synchronization point for units that need to run once during
+              the first boot.  Only after all units ordered before this target have finished, will the
+              machine-id be committed to disk, marking the first boot as completed.  If the boot is aborted
```
maybe make machine-id a man link, i.e.

`citerefentry><refentrytitle>machine-id</refentrytitle><manvolnum>5</manvolnum></citerefentry>`

poettering commented on man/systemd.special.xml:
```diff
@@ -277,6 +278,16 @@
+              the first boot.  Only after all units ordered before this target have finished, will the
+              machine-id be committed to disk, marking the first boot as completed.  If the boot is aborted
+              at any time before that, the next boot will re-run any units with <varname>ConditionFirstBoot=yes</varname>.
+            </para>
```
this should be a "passive" unit, as mentioned earlier. i.e. one that is pulled in by the stutff ordered before it, not pulled in by the stuff ordered after it. There's a special section in the manpage for these, further below "Special Passive System Units"

Rahix replied:
Ah, I misunderstood your earlier comment regarding this.  This makes sense, I'll change it.

poettering commented on units/systemd-machine-id-commit.service:
```diff
@@ -12,8 +12,9 @@ Description=Commit a transient machine-id on disk
-After=local-fs.target
+Before=shutdown.target
+After=local-fs.target first-boot-complete.target
+Wants=first-boot-complete.target
```
no, first-boot-complete.target should be "passive" i think, as mentioned, i.e. pulled in by the stuff ordered before it, not the stuff ordered after it. i.e. please drop the Wants= but keep the After=

poettering commented on units/systemd-firstboot.service:
```diff
@@ -13,7 +13,7 @@ Documentation=man:systemd-firstboot(1)
 Conflicts=shutdown.target
 After=systemd-remount-fs.service
-Before=systemd-sysusers.service sysinit.target shutdown.target
+Before=systemd-sysusers.service sysinit.target first-boot-complete.target shutdown.target
```
this should also pull in the target with Wants=

poettering commented on units/systemd-random-seed.service.in:
```diff
@@ -14,7 +14,7 @@ DefaultDependencies=no
 Conflicts=shutdown.target
 After=systemd-remount-fs.service
-Before=shutdown.target
+Before=first-boot-complete.target shutdown.target
```
ditto

poettering commented:
also, needs rebase

poettering commented:
btw, we'd like to do a release soon. if you post a new version I think we can still merge this before.

Rahix commented:
@poettering, sorry, I got sidetracked by other things.  I've already addressed most open discussions but did not push it yet ... I'll update asap.

Rahix commented:
First push is just addressing the comments ([diff](https://github.com/systemd/systemd/compare/afded2afc8f0c648055d2457f5db3b78393891e3..029820914987847e5f2277d1c653f43072b537a5)), will rebase now as a second step so it is more clear what changed.

Rahix commented:
Rebased and re-tested everything as well now.

Rahix commented:
I don't really understand how this CI error is related to my changes?

```
cp: cannot create regular file '/var/tmp/systemd-test.OyfGFe/minimal/usr/lib/os-release': No such file or directory
make: *** [setup] Error 1
```

poettering commented:
> I don't really understand how this CI error is related to my changes?
> 
> ```
> cp: cannot create regular file '/var/tmp/systemd-test.OyfGFe/minimal/usr/lib/os-release': No such file or directory
> make: *** [setup] Error 1
> ```

I guess #17367 has some explanation for it.

poettering commented on src/core/main.c:
```diff
@@ -2001,15 +2001,26 @@ static void log_execution_mode(bool *ret_first_boot) {
+                        r = read_one_line_file("/etc/machine-id", &id_text);
+                        if (r < 0 || streq(id_text, "uninitialized")) {
+                                if (r < 0 && r != -ENOENT)
+                                        log_warning_errno(r, "Unexpected error while reading /etc/machine-id: %m");
```
we usually add ", ignoring" or a similar string to log messages that indicate errors of some form we ignore. i.e. change the string here to:

```
"Unexpected error while reading /etc/machine-id, ignoring: %m"
```

or so

Rahix replied:
Okay, will do once I'm home this evening :)

Rahix replied:
It's done

poettering commented:
looks excellent. any chance you can tweak that one log message as suggested, too? Should be good to merge then!

Rahix commented:
> any chance you can tweak that one log message as suggested, too?

Done :)

Rahix commented:
Rebased to (hopefully) pull in the fix for CI.

keszybz commented on src/libsystemd/sd-id128/id128-util.c:
```diff
@@ -115,7 +121,7 @@ int id128_read_fd(int fd, Id128Format f, sd_id128_t *ret) {
                 _fallthrough_;
         case 36: /* RFC UUID without trailing newline */
-                if (f == ID128_PLAIN)
+                if (f == ID128_PLAIN || f == ID128_PLAIN_OR_UNINIT)
```
`IN_SET(f, ID128_PLAIN, ID128_PLAIN_OR_UNINIT)`

keszybz commented on man/machine-id.xml:
```diff
@@ -115,6 +116,34 @@
+
+    <orderedlist>
+      <listitem><para>If <filename>/etc/machine-id</filename> does not exist, this is a first boot.  During
+      early boot, systemd will create it with the content <literal>uninitialized</literal> and
```
`unitialized\n` strictly speaking. Not sure if we want to clarify that...

This sentence is awkward. Maybe say:
`During early boot, systemd will write <literal>unitialized\n</literal> to this file and overmount a temporary file which contains the actual machine ID. Later (after <filename>first-boot-complete.target</filename> has been reached), the real machine ID will be written to disk.`

Rahix replied:
> `unitialized\n` strictly speaking. Not sure if we want to clarify that...

Just for reference, the code will also work correctly when `\n` is missing.

keszybz commented on man/machine-id.xml:
```diff
@@ -115,6 +116,34 @@
+      a boot is also considered the first boot.  The same mechanism as above applies.</para></listitem>
+
+      <listitem><para>If <filename>/etc/machine-id</filename> exists and is empty, a boot is
+      <emphasis>not</emphasis> considered the first boot.  systemd will still bind-mount a file containing
```
`<command>systemd</>`

keszybz commented on man/machine-id.xml:
```diff
@@ -115,6 +116,34 @@
+
+      <listitem><para>If <filename>/etc/machine-id</filename> exists and is empty, a boot is
+      <emphasis>not</emphasis> considered the first boot.  systemd will still bind-mount a file containing
+      the actual machine-id over it and later try to commit it to disk (if <filename>/etc</filename> is
```
/etc/

keszybz reviewed (commented):
A few style tweaks please.

Rahix commented:
@keszybz, I've applied your style tweaks.  I've also slightly reworded the documentation about `ConditionFirstBoot` to make it more explicit that the target is a passive one.

yuwata commented:
LGTM.
